Issue #31: Make a facade for logging

yuenj opened the pull request:
This is a simple implementation of a logging facade, but also didn't want to complicate something that was working.  I tried to implement it with the mustache library but it seems to only work with strings, e.g. I can't do Log.note(data) where data is the object returned from the coverage fetch call.

Commit:
Issue #31: Make a facade for logging

klahnakoski commented:
Yes, the first parameter must be a string.  You could have `Log.note(template, params)` perform `template = ""+template;` first, or throw an error if the `template` is not a string.

I understand the appeal of `Log.note(data)` for quick debugging: I personally prefer `Log.note("{{data}}", {data: data})` which is verbose.  You could make a `debug` function that is more flexible for your cause.

```javascript
Log.debug = value => {
    if (DEBUG){
        console.log(""+value)
    }
}
```

Your strategy of keeping it very simple is also just fine.

klahnakoski commented on src/utils/log.js:
```diff
@@ -0,0 +1,11 @@
+export const note = params => console.log(params);
+
+export const info = params => console.info(params);
```
Do you plan to use all of these?  I have only seen two use cases in the code so far:  Logging exceptions and ephemeral debugging logs.  I suggest making only those two for now

klahnakoski replied:
Since logging errors is the major use case, I suggest looking at https://github.com/klahnakoski/firefox-code-coverage-frontend/issues/25 to ensure the API can handle chaining (even if the implementation does nothing).

Commit:
add mustache library

yuenj commented:
I like the idea of using the mustache library but I'm currently running into this issue: https://community.particle.io/t/trying-to-use-mustache-to-return-nested-json-data/24006/6
when I try to Log.note("{{data}}", {data: data}) which returns [object Object],[object Object],[object Object], ... I will try to work something out though.

Thank you for the idea of including the DEBUG flag in the log file itself.

yuenj commented:
Would something like this make sense?
```

export const note = (template, params) => {
  try {
    let template = ''+template;
    if (template === 'undefined') {
      // if no template is specified, do the regular way of logging
      params.forEach((p) => console.log(param));
    } else {
      // a template is specified
      let jsonParams = [];
      params.forEach((p) => {
        const jsonParam = JSON.stringify(p, null, TAB_SIZE);
        jsonParams.push(jsonParam);
      })
      console.log(Mustache.render(template, jsonParams));
    } 
  } catch (error) {
    throw Error(`Problem logging: ${error}`);
  }
}
```

yuenj commented:
Or simplified to this, since mustache takes a object instead of an array: 
```
export const note = (template, params) => {
  try {
    let template = ''+template;
    if (template === 'undefined') {
      // if no template is specified, then do the regular way of logging
      console.log(params);
    } else {
      // a template is specified
      console.log(Mustache.render(template, params));
    } 
  } catch (error) {
    throw Error(`Problem logging: ${error}`);
  }
}
```

klahnakoski commented:
perform the `template === 'undefined'` before you turn it into a string.

klahnakoski commented:
Yes, it is intended that the `params` is an object, not an array; it is a little more verbose, but worth the benefit of having named parameters.

Converting parameters to JSON is a good idea, but runs the risk those parameters are not JSON-serailizable, or self referential, or really large.

Commit:
add template and error checking to logging

yuenj commented:
Updated
1. kept note and error methods, removed the other methods of logging
2. added error checking 
3. debug function is a wrapper for console.log for quick debugging purposes (non-template)

example usage:     
```
Log.debug(data); 
Log.note('{{line}}', { line: covered[0] });
Log.error('{{line}}', { line: covered[0] });
```

yuenj commented:
try: http://localhost:3000/file?revision=b828f7a6c2bf&path=/accessible/atk/Platform.cpp

klahnakoski commented:
Good!  @LinkaiQi please merge if this is ok with you
